auditor-client/src/lib.rs

```diff
@@ -830,6 +899,229 @@ impl AuditorClient {
+/// # use auditor::domain::{RecordAdd, RecordTest};
+/// #
+/// # async fn foo() -> Result<(), ClientError> {
+/// # //let record = RecordAdd::try_from(Faker.fake::<RecordTest>()).unwrap()
+/// # let record = RecordAdd::try_from(RecordTest::default()).unwrap();
+/// let mut client = AuditorClientBuilder::new()
+///     .address(&"localhost", 8000)
+///     .database_path("sqlite://:memory:")
+///     .send_interval(60)
+///     .build_queued()
+///     .await?;
+/// client.add(&record).await?;
+/// client.stop().await?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct QueuedAuditorClient {
+    database: Database,
+    client: AuditorClient,
+    shutdown_tx: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    task_handle: Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>,
+}
+
+impl QueuedAuditorClient {
+    /// Constructs the `QueuedAuditorClient` and starts the background send task
+    fn new(database: Database, client: AuditorClient, interval: std::time::Duration) -> Self {
+        let mut interval = tokio::time::interval(interval);
+        let (shutdown_tx, mut shutdown_rx) = oneshot::channel();
+        let _database = database.clone();
+        let _client = client.clone();
+        // Note: Since the first tick on interval::tick is immediate,
+        // a send is triggered immediately.
+        let task_handle = tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = interval.tick() => {},
+                    result = &mut shutdown_rx => {
+                        if let Err(e) = result { tracing::error!("Error: {:?}", e) }
+                        break;
+                    },
+                }
+                if let Err(e) = Self::process_queue(&_database, &_client).await {
+                    tracing::error!("Processing queue failed with error: {e}");
+                }
+            }
+        });
+        Self {
+            database,
+            client,
+            shutdown_tx: Arc::new(Mutex::new(Some(shutdown_tx))),
+            task_handle: Arc::new(Mutex::new(Some(task_handle))),
+        }
+    }
+
+    #[tracing::instrument(name = "Process client send queue", skip(database, client))]
+    async fn process_queue(database: &Database, client: &AuditorClient) -> Result<(), ClientError> {
+        // Most recent update id
+        let update_rowid = database.get_last_update_rowid().await?;
+
+        // Send all inserts
+        for (rowid, r) in database.get_inserts().await? {
+            match client.add(&r).await {
+                Ok(_) => {
+                    tracing::info!("Successfully sent {} records", r.record_id);
+                    database.delete_insert(rowid).await?;
+                }
+                Err(ClientError::RecordExists) => {
+                    // TODO: make sure auditor actually inserts all new records
+                    tracing::warn!(
+                        "Failed sending record to Auditor instance. Record already exists.",
```

Comment:
Maybe add the record ID to the warning message so that one can better debug this issue.